### PR TITLE
force alpha to 1 when using background color as inverted foreground color

### LINKED ISSUE #1898: Handling transparent BG color when used as FG
Nano displays its shortcuts with FG/BG colors reversed:
image

However, when theme BG is `transparent`, following happens:
image

The solution could be to have a separate flag that will render characters with no background as if background were transparent instead of `theme.backgroundColor`, but will still use the actual color for inverted elements.

### PR DESCRIPTION
**Pull request by ivanwonder**
This problem has been described there #1898.

Fixes #1898

### CODE REVIEW

File: src/browser/renderer/atlas/DynamicCharAtlas.ts [hunk 1]
```diff
@@ -253,7 +254,13 @@ export class DynamicCharAtlas extends BaseCharAtlas {
       `${fontStyle} ${fontWeight} ${this._config.fontSize * this._config.devicePixelRatio}px ${this._config.fontFamily}`;
     this._tmpCtx.textBaseline = 'middle';
 
-    this._tmpCtx.fillStyle = this._getForegroundColor(glyph).css;
+    const fgColor = this._getForegroundColor(glyph);
+    this._tmpCtx.fillStyle = fgColor.css;
+
+    if (glyph.fg === INVERTED_DEFAULT_COLOR) {
+      const rgba = fromRgba(fgColor.rgba);
+      this._tmpCtx.fillStyle = toCss(rgba[0], rgba[1], rgba[2]);
```
Comment (Tyriar) [inline src/browser/renderer/atlas/DynamicCharAtlas.ts]: These transformations are a little more expensive that needed, we could just have a helper that converts an `IColor` (or css or rgba) to be opaque, for example for rgba all that needs to happen is `rgba | 0xFF`.

File: src/browser/renderer/atlas/DynamicCharAtlas.ts [hunk 2]
```diff
@@ -253,7 +254,13 @@ export class DynamicCharAtlas extends BaseCharAtlas {
       `${fontStyle} ${fontWeight} ${this._config.fontSize * this._config.devicePixelRatio}px ${this._config.fontFamily}`;
     this._tmpCtx.textBaseline = 'middle';
 
-    this._tmpCtx.fillStyle = this._getForegroundColor(glyph).css;
+    const fgColor = this._getForegroundColor(glyph);
```
Comment (Tyriar) [inline src/browser/renderer/atlas/DynamicCharAtlas.ts]: For the WebGL renderer it should be a very similar here:

https://github.com/xtermjs/xterm.js/blob/339d4d6a7316d1c1791b7a5859ba9c8175ce3cc6/addons/xterm-addon-webgl/src/atlas/WebglCharAtlas.ts#L209-L211

File: src/browser/renderer/atlas/DynamicCharAtlas.ts (hunk 1 above)
Comment (ivanwonder) [reply src/browser/renderer/atlas/DynamicCharAtlas.ts]: now I add new function `opaque` in Color.ts

**Comment by Eugeny** [comment]
Works with the canvas renderer, but not WebGL. I've poked around a bit but could find why right away

**Comment by Tyriar** [comment]
@Eugeny do you mean transparency in general or this specific issue (`\x1b[7m█\x1b[0m` should be `#ff0000` when the background color is `#ff000080` and allowTransparency is true?

**Comment by Eugeny** [comment]
I meant #1898

**Comment by Tyriar** [comment]
@Eugeny is my test case correct? That was working on all 3 renderers when I tested.

**Comment by Eugeny** [comment]
With a white-on-transparent theme, 

Canvas:
image

WebGL:
image

**Comment by Tyriar** [comment]
@Eugeny what's the background set to exactly?

**Comment by Eugeny** [comment]
Literally `"transparent"`

**Comment by Tyriar** [comment]
@Eugeny oh, that's the problem. I was assuming you were using `#rgba` or `#rrggbbaa`, it should work if you change that to `#00000000`. I hadn't even considered using `transparent` here, that's also why you don't get the scrollbar problem in https://github.com/xtermjs/xterm.js/issues/2252#issuecomment-558371374

**Comment by Tyriar** [comment]
@Eugeny what's your proposal to fix that? It feels like that's sort of undefined behavior and we shouldn't allow using `transparent` for this reason, using 0 for the alpha channel instead lets you specify the opaque inverted version.

**Comment by Eugeny** [comment]
What do you use to convert named colors to RGBA? `transparent` should be equivalent to `rgba(0, 0, 0, 0)` per CSS spec

**Comment by Tyriar** [comment]
@Eugeny 👍 opening issue again to add this case:

https://github.com/xtermjs/xterm.js/blob/8b5c48fe878f830914248e4c1ed78e621deb16c4/addons/xterm-addon-webgl/src/atlas/WebglCharAtlas.ts#L231-L236

**Comment by ivanwonder** [comment]
@Tyriar Because of the `transparent`, so I remove the bg alpha channel by using the RGBA value and then convert it to CSS string. 
https://github.com/ivanwonder/xterm.js/blob/1ff1809ab75be9a9aeee134e6dd5c040f68efda0/src/browser/Color.ts#L59

**Comment by Tyriar** [comment]
@ivanwonder the fix I did for webgl didn't cover when background=transparent though, I'll fix that before we do a release.

**Comment by ivanwonder** [comment]
@Tyriar https://html.spec.whatwg.org/multipage/canvas.html#serialisation-of-a-color
`context.fillStyle` will convert `transparent` to `rgba(0, 0, 0, 0)`.
`if it has alpha equal to 1.0, then the string is a lowercase six-digit hex value, prefixed with a "#" character, Otherwise, the color value has alpha less than 1.0, and the string is the color value in the CSS rgba()`
maybe it will help

**Comment by Tyriar** [comment]
@ivanwonder thanks, good info

**Comment by ivanwonder** [comment]
I just try the idea that using `context.fillStyle`. Looks work fine.
https://github.com/ivanwonder/xterm.js/tree/reverse-transparency

**Pull request merged**